packages/contracts/contracts/L1/tss/TssStakingSlashing.sol

```diff
@@ -0,0 +1,373 @@
+                "invalid pubKey"
+            );
+            deposits[msg.sender].pubKey = _pubKey;
+        }
+
+        // send bit token to staking contract , need user approve first
+        require(
+            IERC20(BitToken).transferFrom(msg.sender, address(this), _amount),
+            "transfer erc20 token failed"
+        );
+        deposits[msg.sender].amount += _amount;
+
+        emit AddDeposit(msg.sender, DepositInfo({ pubKey: _pubKey, amount: _amount }));
+    }
+
+    /**
+     * @notice user who not elected to be validator to withdraw their bit token
+     */
+    function withdrawToken() public nonReentrant {
+        uint256 amount = deposits[msg.sender].amount;
+        require(amount > 0, "do not have deposit");
+        bytes memory pubKey = deposits[msg.sender].pubKey;
+
+        // when not in consensus period or be selected
+        require(
+            !ITssGroupManager(tssGroupContract).memberExistInActive(pubKey) &&
+                !ITssGroupManager(tssGroupContract).memberExistActive(pubKey),
+            "not at the right time"
+        );
+
+        delete deposits[msg.sender];
+
+        require(IERC20(BitToken).transfer(msg.sender, amount), "erc20 transfer failed");
+
+        emit Withdraw(msg.sender, amount);
+    }
+
+    /**
+     * @notice send quit request for the next election
+     */
+    function quit() public nonReentrant {
+        require(deposits[msg.sender].amount > 0, "do not have deposit");
+        // when not in consensus period
+        require(ITssGroupManager(tssGroupContract).inActiveIsEmpty(), "not at the right time");
+        // is active member
+        require(
+            ITssGroupManager(tssGroupContract).memberExistActive(deposits[msg.sender].pubKey),
+            "not at the active group"
+        );
+        for (uint256 i = 0; i < quitList.length; i++) {
+            require(quitList[i] != msg.sender, "already in quitList");
+        }
+        quitList.push(msg.sender);
+    }
+
+    /**
+     * @notice return the quit list
+     */
+    function getQuitList() public view returns (address[] memory) {
+        return quitList;
+    }
+
+    /**
+     * @notice clear the quit list
+     */
+    function clearQuitList() public onlyOwner {
+        delete quitList;
+    }
+
+    /**
+     * @notice verify the slash message then slash
+     * @param _messageBytes the message that abi encode by type SlashMsg
+     * @param _sig the signature of the hash keccak256(_messageBytes)
+     */
+    function slashing(bytes memory _messageBytes, bytes memory _sig) public nonReentrant {
+        SlashMsg memory message = abi.decode(_messageBytes, (SlashMsg));
+        // verify tss member state
+        require(!isJailed(message.jailNode), "the node already jailed");
+        require(
+            ITssGroupManager(tssGroupContract).memberExistActive(deposits[msg.sender].pubKey),
+            "sender not at the active group"
+        );
```

Comment:
There is no use of this judgment in other methods, and this state judgment is not exposed to external use. In order to reduce the complexity of the code, is it better to keep the original method?

Reply:
ok, then make it internal. I still suggest we merge these two judgments into one, for the gas cost consideration, since it will execute storage querying twice.

Reply:
If I combine the two queries into one method, the resulting method execution will still execute the two query methods, perhaps you mean to provide such operations from the underlying tssgroup contract?

Reply:
LGTM